include/dlaf/matrix/matrix_ref.h

```diff
@@ -0,0 +1,175 @@
+    return splitTile(std::move(tile_sender), SubTileSpec{ij_tile, tile_size});
+  }
+
+private:
+  Matrix<const T, D>& mat_const_;
+
+protected:
+  GlobalElementIndex offset_;
+};
+
+template <class T, Device D>
+class MatrixRef : public MatrixRef<const T, D> {
+public:
+  static constexpr Device device = D;
+
+  using ElementType = T;
+  using TileType = Tile<ElementType, D>;
+  using ConstTileType = Tile<const ElementType, D>;
+  using TileDataType = internal::TileData<ElementType, D>;
+  using ReadWriteSenderType = ReadWriteTileSender<T, D>;
+
+  /// Create a sub-matrix of @p mat with an @p offset and @p size.
+  ///
+  /// @param[in] mat is the input matrix,
+  /// @param[in] offset is the offset of the new matrix relative to the input matrix,
+  /// @param[in] size is the size of the new matrix relative to the offset,
+  /// @pre origin.isValid()
+  /// @pre size.isValid()
+  /// @pre origin + size <= mat.size()
+  MatrixRef(Matrix<T, D>& mat, const GlobalElementIndex& offset, const GlobalElementSize& size)
+      : MatrixRef<const T, D>(mat, offset, size), mat_(mat) {}
+
+  // TODO: default, copy, move construction?
+  MatrixRef() = delete;
+
+  /// Returns a sender of the Tile with local index @p index.
+  ///
+  /// @pre index.isIn(distribution().localNrTiles()).
+  ReadWriteSenderType readwrite(const LocalTileIndex& index) noexcept {
+    // Note: this forwards to the overload with GlobalTileIndex which will
+    // handle taking a subtile if needed
+    return readwrite(this->distribution().globalTileIndex(index));
+  }
+
+  /// Returns a sender of the Tile with global index @p index.
+  ///
+  /// @pre the global tile is stored in the current process,
+  /// @pre index.isIn(globalNrTiles()).
+  ReadWriteSenderType readwrite(const GlobalTileIndex& index) {
+    DLAF_ASSERT(index.isIn(this->distribution().nrTiles()), index, this->distribution().nrTiles());
+
+    const auto parent_index(
+        mat_.distribution().globalTileIndexFromSubDistribution(offset_, this->distribution(), index));
+    auto tile_sender = mat_.readwrite(parent_index);
+
+    const auto parent_dist = mat_.distribution();
+    const auto parent_tile_size = parent_dist.tileSize(parent_index);
+    const auto tile_size = this->tileSize(index);
+
+    // If the corresponding tile in the parent distribution is exactly the same
+    // size as the tile in the sub-distribution, we don't need to take a subtile
+    // and can return the tile sender directly. This avoids unnecessary wrapping.
+    if (parent_tile_size == tile_size) {
+      return tile_sender;
+    }
+
+    // Otherwise we have to extract a subtile from the tile in the parent
+    // distribution.
+    const auto ij_tile =
+        parent_dist.tileElementOffsetFromSubDistribution(offset_, this->distribution(), index);
+    return splitTile(std::move(tile_sender), SubTileSpec{ij_tile, tile_size});
+  }
+
+private:
+  Matrix<T, D>& mat_;
+  using MatrixRef<const T, D>::offset_;
+};
+}
```

Comment:
To do: explicitly instantiate?

Reply:
Done in https://github.com/eth-cscs/DLA-Future/pull/934/commits/a8c848f3e176fdf3ac0b2b77022629880b60c795.